returnn/compile.py

```diff
@@ -281,10 +271,44 @@ def run(self):
             "--verbosity",
             "5",
         ]
-        if self.input_names:
-            cmd += ["--input_names", ",".join(self.input_names)]
-        if self.output_names:
-            cmd += ["--output_names", ",".join(self.output_names)]
+
+        # Pass the tensor names here because ReturnnConfig serialization might
+        # potentially reorder via `sort_config=True`, and the order matters for
+        # output tensor <-> tensor name association.
+        input_names = self.input_names or self.collect_tensor_names(
+            str(self.returnn_root), self.returnn_config.config.get("extern_data", {})
+        )
+        output_names = self.output_names or self.collect_tensor_names(
+            str(self.returnn_root), self.returnn_config.config.get("model_outputs", {})
+        )
+        if input_names:
+            cmd += ["--input_names", ",".join(input_names)]
+        if output_names:
+            cmd += ["--output_names", ",".join(output_names)]
 
         util.create_executable("compile.sh", cmd)  # convenience file for manual execution
         sp.run(cmd, check=True)
+
+    @staticmethod
+    def collect_tensor_names(returnn_root: str, data_dict: Dict[str, Any]) -> List[str]:
+        if returnn_root not in sys.path:
+            sys.path.append(returnn_root)
+
+        try:
+            from returnn.tensor import Tensor
+        except ImportError:
+            # For backwards compatibility assume dynamic time axis (i.e. axis 1 in
+            # BTF order) if we cannot import the Tensor class.
+            return [name for k in data_dict.keys() for name in [k, f"{k}:size1"]]
```

Comment:
I don't think the behavior of this function should depend on some import error.
```suggestion
        from returnn.tensor import Tensor
```